File: examples/auction-engine/web2/csharp/auction-engine-api/Controllers/AuctionController.cs
using Microsoft.AspNetCore.Mvc;
using System.Collections.Concurrent;

namespace auction_engine_api.Controllers;

[ApiController]
[Route("api/auctions")]
public class AuctionController : ControllerBase
{
    private static readonly ConcurrentDictionary<string, Auction> _auctions = new();

    public class Auction
    {
        public string ItemId { get; set; } = string.Empty;
        public string Seller { get; set; } = string.Empty;
        public double HighestBid { get; set; }
        public string? HighestBidder { get; set; }
        public bool IsActive { get; set; } = true;
    }

    public class CreateAuctionRequest { public string ItemId { get; set; } = string.Empty; public string Seller { get; set; } = string.Empty; public double StartPrice { get; set; } }
    public class BidRequest { public string Bidder { get; set; } = string.Empty; public double Amount { get; set; } }

    [PostMapping]
    public IActionResult CreateAuction([FromBody] CreateAuctionRequest request)
    {
        var auction = new Auction { ItemId = request.ItemId, Seller = request.Seller, HighestBid = request.StartPrice };
        _auctions[request.ItemId] = auction;
        return Created("", auction);
    }

    [HttpPost("{itemId}/bid")]
    public IActionResult PlaceBid(string itemId, [FromBody] BidRequest request)
    {
        if (!_auctions.TryGetValue(itemId, out var auction) || !auction.IsActive)
            return NotFound("Auction not found");

        if (request.Amount <= auction.HighestBid)
            return BadRequest($"Bid must be higher than {auction.HighestBid}");

        auction.HighestBid = request.Amount;
        auction.HighestBidder = request.Bidder;

        return Ok(new { message = "Bid accepted", highestBid = auction.HighestBid });
    }
}

File: examples/auction-engine/web2/springboot/src/main/java/com/example/demo/AuctionController.java
package com.example.demo;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import java.util.*;
import java.util.concurrent.ConcurrentHashMap;

@RestController
@RequestMapping("/api/auctions")
public class AuctionController {

    private static final Map<String, Auction> auctions = new ConcurrentHashMap<>();

    public static class Auction {
        public String itemId;
        public String seller;
        public double highestBid;
        public String highestBidder;
        public boolean isActive = true;

        public Auction(String itemId, String seller, double startPrice) {
            this.itemId = itemId;
            this.seller = seller;
            this.highestBid = startPrice;
        }
    }

    @PostMapping
    public ResponseEntity<Auction> createAuction(@RequestBody Map<String, Object> payload) {
        String itemId = (String) payload.get("item_id");
        String seller = (String) payload.get("seller");
        double startPrice = Double.parseDouble(payload.get("start_price").toString());

        Auction auction = new Auction(itemId, seller, startPrice);
        auctions.put(itemId, auction);
        return new ResponseEntity<>(auction, HttpStatus.CREATED);
    }

    @PostMapping("/{itemId}/bid")
    public ResponseEntity<?> placeBid(@PathVariable String itemId, @RequestBody Map<String, Object> payload) {
        Auction auction = auctions.get(itemId);
        if (auction == null || !auction.isActive) {
            return ResponseEntity.status(HttpStatus.NOT_FOUND).body("Auction not found");
        }

        double amount = Double.parseDouble(payload.get("amount").toString());
        String bidder = (String) payload.get("bidder");

        if (amount <= auction.highestBid) {
            return ResponseEntity.badRequest().body("Bid must be higher than current: " + auction.highestBid);
        }

        auction.highestBid = amount;
        auction.highestBidder = bidder;
        return ResponseEntity.ok(Map.of("message", "Bid accepted", "highest_bid", amount));
    }
}
